sw/host/opentitanlib/src/image/image.rs

```diff
@@ -160,29 +154,32 @@ impl Image {
         Ok(manifest)
     }
 
-    /// Updates the length field in the `Manifest` to the length of the image.
-    pub fn update_length(&mut self) -> Result<usize> {
-        let length = self.size as u32;
+    /// Updates the length field in the `Manifest`.
+    /// This method requires `unsigned_ext_size` since the `length` field must be updated before
+    /// the image is signed and unsigned fields are typically signature fields.
+    pub fn update_length(&mut self, unsigned_ext_size: usize) -> Result<usize> {
+        let length = (self.size + unsigned_ext_size) as u32;
         let m = self.borrow_manifest_mut()?;
         m.length = length;
         Ok(self.size)
     }
 
-    /// Updates the `signed_reion_end` field
-    pub fn update_signed_region(&mut self) -> Result<usize> {
-        // TODO(#18496): Should depend on extensions.
-        let signed_region_end = self.size as u32;
-        let m = self.borrow_manifest_mut()?;
-        m.signed_region_end = signed_region_end;
-        Ok(self.size)
+    /// Sets the `signed_reion_end` field to `self.size`.
+    /// Note that this method must be called AFTER all signed extensions are added but BEFORE
+    /// adding any unsigned extensions.
+    pub fn update_signed_region_end(&mut self) -> Result<usize> {
+        let signed_region_end = self.size;
+        self.borrow_manifest_mut()?.signed_region_end = signed_region_end as u32;
+        Ok(signed_region_end)
     }
 
     /// Operates on the signed region of the image.
     pub fn map_signed_region<F, R>(&self, f: F) -> R
     where
         F: FnOnce(&[u8]) -> R,
     {
-        f(&self.data.bytes[offset_of!(Manifest, usage_constraints)..self.size])
+        f(&self.data.bytes[offset_of!(Manifest, usage_constraints)
+            ..self.borrow_manifest().unwrap().signed_region_end as usize])
```

Comment:
To avoid using `unwrap()` it would be better to just return `Result<R>` instead.

Reply:
done